### PR TITLE
[Distributed] Offer LocalTestingDistributedActorSystem

### PR DESCRIPTION
**Pull request by yim-lee**
Add local-only actor system `LocalTestingDistributedActorSystem` to make code and tutorial development easier.

rdar://89580224

### CODE REVIEW

**Comment by yim-lee** [comment]
@ktoso Please review

File: stdlib/public/Distributed/DistributedActorSystem.swift [hunk 1]
```diff
@@ -93,7 +93,7 @@ public protocol DistributedActorSystem: Sendable {
   /// and not re-cycled by the system), i.e. if it is called during a failure to initialize completely,
   /// the call from the actor's deinitalizer will not happen (as under these circumstances, `deinit` will be run).
   ///
-  /// If the `actorReady` gets called with some unknown ID, it should crash immediately as it signifies some
+  /// If `resignID` gets called with some unknown ID, it should crash immediately as it signifies some
```
Comment (ktoso) [inline stdlib/public/Distributed/DistributedActorSystem.swift:96]: right thanks!

**Comment by yim-lee** [comment]
@swift-ci please smoke test

File: stdlib/public/Distributed/LocalTestingDistributedActorSystem.swift [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+// TODO(distributed): not thread safe...
+public final class LocalTestingDistributedActorSystem: DistributedActorSystem, @unchecked Sendable {
+  public typealias ActorID = ActorAddress
+  public typealias InvocationEncoder = LocalInvocationEncoder
+  public typealias InvocationDecoder = LocalInvocationDecoder
```
Comment (ktoso) [inline stdlib/public/Distributed/LocalTestingDistributedActorSystem.swift:17]: Maybe `LocalTesting...`? Just to have the same full prefix hm

File: stdlib/public/Distributed/LocalTestingDistributedActorSystem.swift [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    guard let actor = anyActor as? Act else {
+      throw LocalTestingDistributedActorSystemError(message: "Failed to resolve id '\(id)' as \(Act.Type)")
+    }
+    return actor
+  }
+
+  public func assignID<Act>(_ actorType: Act.Type) -> ActorID
+    where Act: DistributedActor {
+    let id = ActorAddress(parse: "\(self.idProvider.next())")
+    print("| assign id: \(id) for \(actorType)")
```
Comment (ktoso) [inline stdlib/public/Distributed/LocalTestingDistributedActorSystem.swift:41]: we can drop all the printlns :)

File: stdlib/public/Distributed/LocalTestingDistributedActorSystem.swift [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  public func resolve<Act>(id: ActorID, as actorType: Act.Type)
+    throws -> Act? where Act: DistributedActor {
+    guard let anyActor = self.activeActors[id] else {
+      throw LocalTestingDistributedActorSystemError(message: "Unable to locate id '\(id)' locally")
+    }
+    guard let actor = anyActor as? Act else {
+      throw LocalTestingDistributedActorSystemError(message: "Failed to resolve id '\(id)' as \(Act.Type)")
+    }
+    return actor
+  }
```
Comment (ktoso) [inline stdlib/public/Distributed/LocalTestingDistributedActorSystem.swift:36]: 👍

**Review by ktoso — COMMENTED** [review]
Looking good :) Need to figure out what locks we can use here to make it threadsafe...

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Review by ktoso — COMMENTED** [review]
Looks good! We need to make it thread safe and we're good here :)

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test Linux

**Comment by ktoso** [comment]
https://github.com/apple/swift/pull/41804 was fixed so we can now use the `any` spelling

**Comment by yim-lee** [comment]
@swift-ci please smoke test

File: stdlib/public/Distributed/LocalTestingDistributedActorSystem.swift [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+  public let message: String
+
+  public init(message: String) {
+    self.message = message
+  }
+}
+
+// === lock ----------------------------------------------------------------
+
+fileprivate class _Lock {
```
Comment (ktoso) [inline stdlib/public/Distributed/LocalTestingDistributedActorSystem.swift:200]:
```suggestion
@available(SwiftStdlib 5.7, *)
fileprivate class _Lock {
```

should help fix the mac CI

**Comment by ktoso** [comment]
Not sure why I can't PR into your branch, but here's a minor patch that'll make us able ot merge this https://github.com/ktoso/swift/commit/a49310b8415f8e2a9e8848e29983e96704ee30f3

Thanks a lot @yim-lee !

**Comment by yim-lee** [comment]
Thanks @ktoso. cherry-picked.

@swift-ci please smoke test

File: test/Distributed/Runtime/distributed_actor_localSystem.swift [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+// REQUIRES: executable_test
+// REQUIRES: concurrency
+// REQUIRES: distributed
+
+// rdar://76038845
+// UNSUPPORTED: use_os_stdlib
+// UNSUPPORTED: back_deployment_runtime
+
+import _Distributed
```
Comment (ktoso) [inline test/Distributed/Runtime/distributed_actor_localSystem.swift:13]: sadly gotta test this all again since module name change

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge

**Pull request merged**